### PR TITLE
KM182 ✅ Add ECR provisioning

### PR DESCRIPTION
**Pull request by deifyed**
We will merge this PR after #426 has been merged into master.

## Description

Adds functionality to create an ECR repository with CloudFormation on `apply application`.

Still missing:
* Delete ECR repository (missing UX plan)
* Possibly functionality for accessing ECR from the EKS cluster
* UX for using an ECR repo in the `app.yaml`
* Documentation

## Motivation and Context

[KM182](https://trello.com/c/KYvl0GiF)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have updated the release notes (for the next release).

### CODE REVIEW

**Review by yngvark — COMMENTED** [review]
This is not a complete review, I'm going to come back to this after focusing on Nightly.

File: cmd/okctl/apply_application.go [hunk 1]
```diff
@@ -99,11 +115,60 @@ func buildApplyApplicationCommand(o *okctl.Okctl) *cobra.Command {
...
 	}
 
 	cmd.Flags().StringVarP(&opts.File, "file", "f", "", "Specify the file path. Use \"-\" for stdin")
 
 	return cmd
 }
+
+func writeSuccessMessage(writer io.Writer, applicationName, argoCDResourcePath string) error {
+	templateString := `
+	Successfully scaffolded {{ .ApplicationName }}
```
[the remark this answers is not in the document]
Comment (paulbes) [reply cmd/okctl/apply_application.go:146]: `applucation`, almost a new unintentional band name

File: pkg/api/container_repository_api.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"context"
+
+	validation "github.com/go-ozzo/ozzo-validation/v4"
+)
+
+// CreateContainerRepositoryOpts contains necesessary information to create a container repository
+type CreateContainerRepositoryOpts struct {
+	ClusterID ID
```
Comment (paulbes) [inline pkg/api/container_repository_api.go:11]: We call this `ID` everywhere else, super minor detail, I know, but might as well stick with it

File: pkg/cfn/components/containerrepository/containerrepository.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+// ContainerRepository contains state for building a cloud formation resource
+type ContainerRepository struct {
+	clusterName string
+	environment string
+	imageName   string
+}
+
+// Resource returns the cloud formation resource
+func (c *ContainerRepository) Resource() cloudformation.Resource {
+	return &ecr.Repository{
```
Comment (paulbes) [inline pkg/cfn/components/containerrepository/containerrepository.go:43]: Love these defaults ❤️

File: pkg/client/application_client.go [hunk 4]
```diff
@@ -67,14 +75,15 @@ func (o OkctlApplication) HasService() bool {
...
+		validation.Field(&o.Namespace, validation.Required, is.Subdomain),
+		validation.Field(&o.Image, validation.Required),
 		validation.Field(&o.Version, validation.Required),
-		validation.Field(&o.ImagePullSecret, validation.Required, is.DNSName),
+		validation.Field(&o.ImagePullSecret, is.Subdomain),
 		validation.Field(&o.SubDomain, is.Subdomain),
-		validation.Field(&o.Port, is.Port),
-		validation.Field(&o.Replicas, validation.Min(0)),
+		validation.Field(&o.Port, validation.Min(lowestPossiblePort), validation.Max(highestPossiblePort)),
+		validation.Field(&o.Replicas, validation.Min(lowestPossibleReplicas)),
```
Comment (paulbes) [inline pkg/client/application_client.go:85]: I look forward to creating a deployment with 34567898765434567 replicas 😬 (just trying to be funny, don't change it)

File: pkg/client/container_repository_client.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	DeleteContainerRepository(ctx context.Context, opts DeleteContainerRepositoryOpts) error
+}
+
+// ContainerRepositoryAPI invokes the API
+type ContainerRepositoryAPI interface {
+	CreateContainerRepository(opts api.CreateContainerRepositoryOpts) (*api.ContainerRepository, error)
+	DeleteContainerRepository(opts api.DeleteContainerRepositoryOpts) error
+}
+
+// ContainerRepositoryStore saves the data
```
Comment (paulbes) [inline pkg/client/container_repository_client.go:44]: You can keep it if you want to, but I will remove it in my state PR, so if you waaaant to 😬

File: pkg/client/container_repository_client.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+// ContainerRepositoryState updates the state
+type ContainerRepositoryState interface {
+	SaveContainerRepository(repository *ContainerRepository) (*store.Report, error)
+	RemoveContainerRepository(imageName string) (*store.Report, error)
+	GetContainerRepository(imageName string) (*ContainerRepository, error)
+}
+
+// ContainerRepositoryReport reports on the state and storage operations
+type ContainerRepositoryReport interface {
```
Comment (paulbes) [inline pkg/client/container_repository_client.go:58]: You can keep it if you want to, but I will remove it in my state PR, so if you waaaant to 😬

File: pkg/client/core/report/console/container_repository_console.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
+package console
```
Comment (paulbes) [inline pkg/client/core/report/console/container_repository_console.go:1]: You can keep it if you want to, but I will remove it in my state PR, so if you waaaant to 😬

File: pkg/client/core/service_container_repository_client.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	store   client.ContainerRepositoryStore
+	state   client.ContainerRepositoryState
+	report  client.ContainerRepositoryReport
+
+	provider v1alpha1.CloudProvider
+}
+
+// CreateContainerRepository handles api, state, store and report orchistration for creation of a container repository
+func (c *containerRepositoryService) CreateContainerRepository(_ context.Context, opts client.CreateContainerRepositoryOpts) (*client.ContainerRepository, error) {
+	err := c.spinner.Start("container registry")
```
Comment (paulbes) [inline pkg/client/core/service_container_repository_client.go:29]: You can keep it if you want to, but I will remove it in my state PR, so if you waaaant to 😬 (the spinner)

File: pkg/client/core/store/filesystem/container_repository_filesystem.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
+package filesystem
```
Comment (paulbes) [inline pkg/client/core/store/filesystem/container_repository_filesystem.go:1]: You can keep it if you want to, but I will remove it in my state PR, so if you waaaant to 😬

**Review by paulbes — APPROVED** [review]
This is the first time you pull something through the whole stack on your own I believe? Absolutely gorgeous! Great job ❤️

File: pkg/api/container_repository_api.go (hunk 2 above)
Comment (deifyed) [reply pkg/api/container_repository_api.go:11]: Discussed in DM and agreed upon renaming the api.ID struct to api.ClusterID

File: pkg/client/application_client.go (hunk 4 above)
Comment (deifyed) [reply pkg/client/application_client.go:85]: yeah! left it open in case one of the big cloud providers decides to migrate to okctl

File: pkg/client/container_repository_client.go (hunk 5 above)
Comment (deifyed) [reply pkg/client/container_repository_client.go:44]: 👍

File: cmd/okctl/apply_application.go [hunk 10]
```diff
@@ -99,11 +115,60 @@ func buildApplyApplicationCommand(o *okctl.Okctl) *cobra.Command {
...
 	return cmd
 }
+
+func writeSuccessMessage(writer io.Writer, applicationName, argoCDResourcePath string) error {
+	templateString := `
+	Successfully applied {{ .ApplicationName }}
+	To deploy your application:
+		1. Commit and push the changes done by okctl
+		2. Run kubectl apply -f {{ .ArgoCDResourcePath }}
+	If using an ingress, it can take up to five minutes for the routing to configure
```
Comment (yngvark) [inline cmd/okctl/apply_application.go:150]: Haven't we seen that it can take 10-12 minutes to setup a new ingress?

File: pkg/cfn/components/composer.go [hunk 11]
```diff
@@ -1630,3 +1632,34 @@ func NewS3BucketComposer(bucketName, repo, env string) *S3BucketComposer {
...
+// ECRRepositoryComposer contains state required for creating a ECR repository
+type ECRRepositoryComposer struct {
+	ImageName   string
+	Repository  string
+	Environment string
+}
+
+// ResourceRepositoryNameOutput returns the name of the resource
+func (e *ECRRepositoryComposer) ResourceRepositoryNameOutput() string {
+	return fmt.Sprintf("%s%s%sECRepository", e.Repository, e.Environment, e.ImageName)
```
Comment (yngvark) [inline pkg/cfn/components/composer.go:1645]: ECRepository -> ECRRepository

File: pkg/cfn/components/composer.go [hunk 12]
```diff
@@ -1630,3 +1632,34 @@ func NewS3BucketComposer(bucketName, repo, env string) *S3BucketComposer {
...
+	repository := containerrepository.New(e.Repository, e.Environment, e.ImageName)
+
+	return &cfn.Composition{
+		Outputs:   []cfn.StackOutputer{repository},
+		Resources: []cfn.ResourceNamer{repository},
+	}, nil
+}
+
+// NewECRepositoryComposer returns an initialized ECR composer
+func NewECRepositoryComposer(imageName, repository, environment string) *ECRRepositoryComposer {
```
Comment (yngvark) [inline pkg/cfn/components/composer.go:1659]: NewECRepositoryComposer -> NewECRRepositoryComposer

**Review by yngvark — APPROVED** [review]
Looks good :+1:

File: cmd/okctl/apply_application.go (hunk 10 above)
Comment (deifyed) [reply cmd/okctl/apply_application.go:150]: Yepp. Old message. I'll update

File: pkg/cfn/components/composer.go (hunk 11 above)
Comment (deifyed) [reply pkg/cfn/components/composer.go:1645]: Since both you and @paulbes agreed on this, I'll change it. But you guys do realize that results in "Elastic Container Registry Repository" right?

File: pkg/config/state/repository.go [hunk 13]
```diff
@@ -89,6 +89,7 @@ type Cluster struct {
 	IdentityPool IdentityPool
 	Monitoring   Monitoring
 	Databases    map[string]Database
+	Repositories map[string]ContainerRepository
```
[the remark this answers is not in the document]
Comment (deifyed) [reply pkg/config/state/repository.go:92]: @paulbes removes this file with his major cleanup PR, so I'm choosing to ignore this :D

File: pkg/cfn/components/composer.go (hunk 11 above)
Comment (yngvark) [reply pkg/cfn/components/composer.go]: But isn't that what it is called? The AWS [user guide](https://docs.aws.amazon.com/AmazonECR/latest/userguide/Repositories.html) title is "Amazon ECR private repositories".

File: cmd/okctl/apply_application.go [hunk 14]
```diff
@@ -42,10 +42,6 @@ func buildApplyApplicationCommand(o *okctl.Okctl) *cobra.Command {
 		Args:  cobra.ExactArgs(requiredApplyApplicationArguments),
 		PreRunE: func(cmd *cobra.Command, args []string) error {
 			err := o.Initialise()
-			if err != nil {
```
Comment (yngvark) [inline cmd/okctl/apply_application.go:45]: Why is this removed?

**Review by yngvark — APPROVED** [review]
Splendid!

File: cmd/okctl/apply_application.go (hunk 14 above)
Comment (paulbes) [reply cmd/okctl/apply_application.go:45]: Oh, no, it shouldnt be, probably a mistake from the merging. Strange that wasmt caught by linting

Comment (deifyed) [reply cmd/okctl/apply_application.go]: I fixded
